mainloop.lua

```diff
@@ -218,22 +208,296 @@ do
+    current_use_music_from = percent == 1 and "stage" or "characters"
+  end
+end
+
+function Stack.wait_for_random_character(self)
+  if self.character == random_character_special_value then
+    self.character = uniformly(characters_ids_for_current_theme)
+  elseif characters[self.character]:is_bundle() then -- may have picked a bundle
+    self.character = uniformly(characters[self.character].sub_characters)
+  end
+  character_loader_load(self.character)
+  character_loader_wait()
+end
+
+local function commonGameSetup()
+  stop_the_music()
+  use_current_stage()
+  pick_use_music_from()
+end
+
+function createNewReplay(mode)
+  local result = {}
+  result.engineVersion = VERSION
+  result.replayVersion = 2
+
+  result[mode] = {}
+  local modeReplay = result[mode]
+
+  if mode == "endless" or mode == "time" then
+    modeReplay.do_countdown = P1.do_countdown or false
+    modeReplay.speed = P1.speed
+    modeReplay.difficulty = P1.difficulty
+    modeReplay.cur_wait_time = P1.cur_wait_time or default_input_repeat_delay
+    modeReplay.in_buf = ""
+  elseif mode == "vs" then
+    modeReplay.P = ""
+    modeReplay.O = ""
+    modeReplay.I = ""
+    modeReplay.Q = ""
+    modeReplay.in_buf = ""
+    modeReplay.P1_level = P1.level
+    modeReplay.P1_name = GAME.battleRoom.playerNames[1]
+    modeReplay.P1_char = P1.character
+    modeReplay.P1_char = P1.character
+    modeReplay.P1_cur_wait_time = P1.cur_wait_time
+    modeReplay.do_countdown = true
+    if P2 then
+      modeReplay.P2_level = P2.level
+      modeReplay.P2_name = GAME.battleRoom.playerNames[2]
+      modeReplay.P2_char = P2.character
+      modeReplay.P2_cur_wait_time = P2.cur_wait_time
+
+      modeReplay.P1_win_count = GAME.match.battleRoom.playerWinCounts[1]
+      modeReplay.P2_win_count = GAME.match.battleRoom.playerWinCounts[2]
+    end
+  end
+
+  return result
+end
+
+local function handle_pause(self)
+  if GAME.match.supportsPause then
+    if menu_pause() then
+      GAME.gameIsPaused = not GAME.gameIsPaused
+
+      setMusicPaused(GAME.gameIsPaused)
+
+      if not GAME.renderDuringPause then
+        if GAME.gameIsPaused then
+          reset_filters()
+        else
+          use_current_stage()
+        end
+      end
+    end
+  end
+end
+
+local function finalizeAndWriteReplay(extraPath, extraFilename)
+
+  replay[GAME.match.mode].in_buf = P1.input_buffer_record
+  replay[GAME.match.mode].P = P1.panel_buffer_record
+  replay[GAME.match.mode].Q = P1.gpanel_buffer_record
+
+  local now = os.date("*t", to_UTC(os.time()))
+  local sep = "/"
+  local path = "replays" .. sep .. "v" .. VERSION .. sep .. string.format("%04d" .. sep .. "%02d" .. sep .. "%02d", now.year, now.month, now.day)
+  if extraPath then
+    path = path .. sep .. extraPath
+  end
+  local filename = "v" .. VERSION .. "-" .. string.format("%04d-%02d-%02d-%02d-%02d-%02d", now.year, now.month, now.day, now.hour, now.min, now.sec)
+  if extraFilename then
+    filename = filename .. extraFilename
```

Comment:
Adding some separation
```suggestion
    filename = filename .. "-" .. extraFilename
```